feat: useExportsLast

tommymorgan opened the pull request:
<!--
	Thanks for submitting a Pull Request! We appreciate you spending the time to work on these changes.
	Please provide enough information so that others can review your PR.
	Once created, your PR will be automatically labeled according to changed files.
	Learn more about contributing: https://github.com/biomejs/biome/blob/main/CONTRIBUTING.md
-->

## Summary

<!-- Explain the **motivation** for making this change. What existing problem does the pull request solve?-->
This implements similar functionality to [eslint-plugin-import/exports-last](https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/exports-last.md)

<!-- Link any relevant issues if necessary or include a transcript of any Discord discussion. -->

## Test Plan

Added tests and snapshots

tommymorgan commented:
`valid_multiple_separate_exports.js` uses code taken directly from one of the examples at https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/exports-last.md.  That code is MIT licensed and I want to provide proper attribution and could use a little guidance to make sure I get it right.

tommymorgan commented on crates/biome_js_analyze/tests/specs/nursery/useExportsLast/valid_multiple_separate_exports.js:
```diff
@@ -0,0 +1,11 @@
+const arr = ["bar"];
+
+export const bool = true;
+
+export default bool;
+
+export function func() {
+	console.log("Hello World 🌍");
+}
+
+export const str = "foo";
```
This example code was taken directly from one of the examples at https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/exports-last.md.  That code is MIT licensed and I want to provide proper attribution and could use a little guidance to make sure I get it right.

togami2864 replied:
No problem to add some modification

togami2864 commented on crates/biome_js_analyze/src/lint/nursery/use_exports_last.rs:
```diff
@@ -0,0 +1,80 @@
+
+impl Rule for UseExportsLast {
+    type Query = Ast<JsModuleItemList>;
+    type State = JsSyntaxNode;
```
I think `TextRange` is enough
```suggestion
    type State = TextRange;
```

togami2864 commented on crates/biome_js_analyze/src/lint/nursery/use_exports_last.rs:
```diff
@@ -0,0 +1,80 @@
+
+        for item in items.iter() {
+            if matches!(item, AnyJsModuleItem::JsExport(_)) {
+                last_export = Some(item.syntax().clone());
```
```suggestion
                last_export = Some(item.range());
```

togami2864 commented on crates/biome_js_analyze/src/lint/nursery/use_exports_last.rs:
```diff
@@ -0,0 +1,80 @@
+        None
+    }
+
+    fn diagnostic(_ctx: &RuleContext<Self>, node: &Self::State) -> Option<RuleDiagnostic> {
+        Some(
+            RuleDiagnostic::new(
+                rule_category!(),
+                node.text_range(),
```
```suggestion
    fn diagnostic(_ctx: &RuleContext<Self>, range: &Self::State) -> Option<RuleDiagnostic> {
        Some(
            RuleDiagnostic::new(
                rule_category!(),
                range,
```

togami2864 commented on crates/biome_js_analyze/tests/specs/nursery/useExportsLast/invalid.js:
```diff
@@ -0,0 +1,3 @@
+export const a = 1;
```
Could you check the original test file to add some more cases?
https://github.com/import-js/eslint-plugin-import/blob/main/tests/src/rules/exports-last.js

togami2864 reviewed (commented):
Thank you! I left some comments

tommymorgan commented:
@togami2864 Thank you so much for the feedback.  I'm pretty new to rust and biome and I appreciate the guidance.  I think I've addressed all your concerns and I also resolved the merge conflicts including re-running `just gen-lint` to resolve conflicts with generated files.

togami2864 reviewed (approved):
Thank you! Could you rebase?

ematipico commented on crates/biome_js_analyze/src/lint/nursery/use_exports_last.rs:
```diff
@@ -0,0 +1,79 @@
+use biome_rowan::{AstNode, AstNodeList, TextRange};
+
+declare_lint_rule! {
+    /// Require that all exports are declared after all non-export statements.
+    ///
+    /// Enforces that export statements are placed at the end of the module, after all other statements.
+    ///
+    /// Source: https://github.com/import-js/eslint-plugin-import/blob/main/docs/rules/exports-last.md
```
```suggestion
    /// Require that all exports are declared after all non-export statements.
    ///
    /// Enforces that export statements are placed at the end of the module, after all other statements.
    ///
```

Not needed

ematipico commented on crates/biome_js_analyze/src/lint/nursery/use_exports_last.rs:
```diff
@@ -0,0 +1,79 @@
+            RuleDiagnostic::new(
+                rule_category!(),
+                range,
+                markup! {
+                    "Export statements should appear at the end of the file."
+                },
+            )
+            .note(markup! {
+                "Move this export to the end of the file."
+            }),
```
In our contribution guide, we explain how to [explain a diagnostic to the user](https://github.com/biomejs/biome/blob/main/crates/biome_analyze/CONTRIBUTING.md#explain-a-rule-to-the-user). In this case, we don't explain the error. Can you address that, please?

tommymorgan replied:
I made an update that I hope is acceptable.  I had previously read the contribution guide but I just missed the mark.  Thank you for providing such detailed docs though.  That and the PR help has really made making my first constribution approachable.

ematipico replied:
I believe there's been a misunderstanding. The previous message was fine, even better than the new one. What was missing is _explain why exports should be moved_, because that's the error. So the diagnostic should provide three notes:
1. Error
2. Explain the error 
3. Provide an actionable solution
